model/convert/service_event.go

```diff
@@ -883,6 +893,90 @@ func invalidCadenceTypeError(
 	)
 }
 
+// convertServiceEventProtocolStateVersionUpgrade converts a Cadence instance of the VersionBeacon
+// service event to the protocol-internal representation.
+// CAUTION: This function must only be used for input events computed locally, by an
+// Execution or Verification Node; it is not resilient to malicious inputs.
+// No errors are expected during normal operation.
+func convertServiceEventProtocolStateVersionUpgrade(event flow.Event) (*flow.ServiceEvent, error) {
+	payload, err := ccf.Decode(nil, event.Payload)
+	if err != nil {
+		return nil, fmt.Errorf("could not unmarshal event payload: %w", err)
+	}
+
+	versionUpgrade, err := DecodeCadenceValue("ProtocolStateVersionUpgrade payload", payload,
+		func(cdcEvent cadence.Event) (flow.ProtocolStateVersionUpgrade, error) {
```

Comment:
Can we simplify it a bit by returning a pointer? In the end we still convert to pointer. Not sure if the api of `DecodeCadenceValue` allows it though.

Reply:
Good point - we can return pointer directly 👍. Done in https://github.com/onflow/flow-go/pull/5428/commits/ec6e8781b3860268f3ba46c71e73b6aea6f21c55.